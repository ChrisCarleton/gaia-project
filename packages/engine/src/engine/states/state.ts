export interface State {
  // Player actions
  buildMine(): void;
  startGaiaProject(): void;
  upgradeStructure(): void;
  formFederation(): void;
  advanceResearch(): void;
  powerOrQicAction(): void;
  specialAction(): void;
  freeAction(): void;
  pass(): void;

  // Await player decision
  waitForPlayer(): void;

  // Maintenance
  doIncome(): void;
  doRoundCleanup(): void;
  doEndGameScoring(): void;

  // Player stuff
  addPlayer(): void;
  removePlayer(): void;
}
