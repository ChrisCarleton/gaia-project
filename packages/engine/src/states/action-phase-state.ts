import { SerializedState } from '../core/serialization';
import { EventType, ObserverPublisher } from '../events';
import {
  ChangeStateFunction,
  GameAction,
  GameContext,
  GameState,
  MapHex,
  Player,
  ResearchArea,
} from '../interfaces';
import { StateBase } from './state-base';

export class ActionPhaseState extends StateBase {
  constructor(
    context: GameContext,
    events: ObserverPublisher,
    changeState: ChangeStateFunction,
    private readonly player: Player,
  ) {
    super(context, events, changeState);
  }

  get currentState(): GameState {
    return GameState.ActionPhase;
  }

  init(): void {
    this.events.publish({
      type: EventType.AwaitingPlayerInput,
      allowedActions: [
        GameAction.BuildMine,
        GameAction.FormFederation,
        GameAction.Free,
        GameAction.GaiaProject,
        GameAction.Pass,
        GameAction.PowerOrQic,
        GameAction.Research,
        GameAction.Special,
        GameAction.UpgradeStructure,
      ],
      gameState: this.currentState,
      player: this.player,
    });
  }

  advanceResearch(area: ResearchArea): void {}

  buildMine(location: MapHex): void {
    /*
      TODO: Validate move.
        1. Player must have a mine available in their supply.
        2. Selected hex must have an uninhabited planet.
          a. Exception: Lantids can build mines on other player's planets - they don't pay terraforming costs.
        3. Planet must be within range of one of the player's other planets. TODO: Create a utility function to calculate range.
          a. Base range is determined by the player's progress on the Navigation research track.
          b. Base range can be extended by 2 by spending a QIC.
        4. Planet must be habitable (or it must be capable of being made habitable by terraforming).
          a. Gaia planets can be made habitable by spending 1 QIC (instead of normal cost)
          b. Transdim planets that have been converted to Gaia planets and still have a Gaiaformer in orbit can have mines built on them for the normal cost.
        5. Player must have 1 ore and 2 credits to spend + any additional terraforming costs.
    */
    /*
      TODO:
        1. Player pays resources.
        2. Gaiaformer is returned (if Gaia project).
        3. Mine is placed and event is published.
        4. Player gains 2 knowledge if all of these conditions are met:
          a. Player is playing as Lantids
          b. They have built their planetary institute
          c. The new mine that was just built was built on a planet occupied by another player,
    */
    // TODO: Next action!
  }

  startGaiaProject(location: MapHex): void {
    /*
    TODO: Validate move.
      1. Player must have an available Gaiaformer.
      2. Player must have access to a transdim planet.
        a. An unclaimed transdim planet must be within range of one of the player's other planets.
        b. Players can activate QICs to extend their range.
      3. Player must have enough power nodes available. (See Gaiaforming research track.)
    */
    /*
    TODO: Perform action.
      1. Power nodes are moved to the Gaia area of the power cycle.
      2. Any QICs used to extend range are used up.
      3. Gaiaformer is placed on the trandim planet.
    */
  }

  toJSON(): SerializedState {
    return {
      type: GameState.ActionPhase,
      player: this.context.players.findIndex(
        (player) => player.id === this.player.id,
      ),
    };
  }
}
