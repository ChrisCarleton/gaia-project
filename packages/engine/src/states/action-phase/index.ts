export * from './free-actions';
export * from './pass-action';
export * from './research-action';

export * from './determine-next-state';
