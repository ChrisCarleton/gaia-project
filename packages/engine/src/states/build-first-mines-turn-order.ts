import { FactionType, Player } from '../interfaces';

// TODO: Remove Partial when done.
type TurnOrderStrategies = Record<
  BuildFirstMinesPass,
  (
    players: Readonly<Player[]>,
    currentIndex: number,
  ) => BuildFirstMinesOptions | null
>;

export enum BuildFirstMinesPass {
  First,
  Second,
  Xenos, // Xenos get to build a third mine after everyone else has built two.
  Ivits, // Ivits do not build mines. Instead, they wait until everyone else has gone and then build their planetary institute.
}

export type BuildFirstMinesOptions = {
  playerIndex: number;
  pass: BuildFirstMinesPass;
};

export class BuildFirstMinesTurnOrder {
  private static strategies: TurnOrderStrategies = {
    [BuildFirstMinesPass.First]: (players, currentIndex) => {
      let nextIndex = currentIndex + 1;

      if (players[nextIndex]?.faction.factionType === FactionType.Ivits) {
        nextIndex++;
      }

      if (nextIndex >= players.length) {
        nextIndex = players.length - 1;

        if (players[nextIndex].faction.factionType === FactionType.Ivits) {
          nextIndex--;
        }

        return {
          playerIndex: nextIndex,
          pass: BuildFirstMinesPass.Second,
        };
      }

      return {
        playerIndex: nextIndex,
        pass: BuildFirstMinesPass.First,
      };
    },

    [BuildFirstMinesPass.Second]: (players, currentIndex) => {
      let nextIndex = currentIndex - 1;

      if (nextIndex < 0) {
        // If a player is playing as the Xenos, then they are allowed to place a third mine after everyone has placed their second.
        nextIndex = players.findIndex(
          (player) => player.faction.factionType === FactionType.Xenos,
        );
        if (nextIndex > -1) {
          return {
            playerIndex: nextIndex,
            pass: BuildFirstMinesPass.Xenos,
          };
        }

        // If a player is playing as the Ivits, then they can now place their Planetary Institute.
        nextIndex = players.findIndex(
          (player) => player.faction.factionType === FactionType.Ivits,
        );
        if (nextIndex > -1) {
          return {
            playerIndex: nextIndex,
            pass: BuildFirstMinesPass.Ivits,
          };
        }

        return null;
      }

      return {
        pass: BuildFirstMinesPass.Second,
        playerIndex: nextIndex,
      };
    },

    [BuildFirstMinesPass.Xenos]: (players) => {
      // If a player is playing as the Ivits, then they can now place their Planetary Institute.
      const nextIndex = players.findIndex(
        (player) => player.faction.factionType === FactionType.Ivits,
      );

      if (nextIndex > -1) {
        return {
          playerIndex: nextIndex,
          pass: BuildFirstMinesPass.Ivits,
        };
      }

      return null;
    },

    [BuildFirstMinesPass.Ivits]: () => null,
  };

  static determineNextState(
    pass: BuildFirstMinesPass,
    players: Readonly<Player[]>,
    currentIndex: number,
  ): BuildFirstMinesOptions | null {
    return BuildFirstMinesTurnOrder.strategies[pass](players, currentIndex);
  }
}
