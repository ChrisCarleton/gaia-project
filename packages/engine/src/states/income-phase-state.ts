import {
  ChangeStateFunction,
  EventType,
  GameContext,
  GameState,
  Income,
  Observer,
  RoundBoosterBonusType,
  StructureType,
} from '..';
import { SerializedState } from '../core/serialization';
import { GameCompletedState } from './game-completed-state';
import { StateBase } from './state-base';

export class IncomePhaseState extends StateBase {
  constructor(
    context: GameContext,
    events: Observer,
    changeState: ChangeStateFunction,
  ) {
    super(context, events, changeState);
  }

  get currentState(): GameState {
    return GameState.IncomePhase;
  }

  init(): void {
    this.context.players.forEach((player) => {
      const income: Income[] = [];

      /*
        TODO: Perform income calculations for each player.
          * ✅ Income from structures
          * ✅ Income from round booster
          * Income from research
          * Income from tech tiles
      */

      // Income from structures.
      const factionIncome = player.faction.income;
      income.push(
        factionIncome[StructureType.Mine][player.structures.mine.active],
        factionIncome[StructureType.TradingStation][
          player.structures.tradingStation.active
        ],
        factionIncome[StructureType.ResearchLab][
          player.structures.researchLab.active
        ],
        factionIncome[StructureType.PlanetaryInstitute][
          player.structures.planetaryInstitute.active
        ],
      );
      // TODO: Check for acadamy income.

      // Round booster income.
      if (player.roundBooster) {
        const { a, b } = player.roundBooster;

        if (a.type === RoundBoosterBonusType.Income) {
          income.push(a.income);
        }

        if (b.type === RoundBoosterBonusType.Income) {
          income.push(b.income);
        }
      }

      this.events.publish({
        type: EventType.IncomeGained,
        player,
        income: income.reduce(this.reduceIncome),
      });
    });

    // TODO: Change state.
    this.changeState(
      new GameCompletedState(this.context, this.events, this.changeState),
    );
  }

  private reduceIncome(total: Income, currrentValue: Income): Income {
    return {
      ...{
        chargePower: this.addResource(
          total.chargePower,
          currrentValue.chargePower,
        ),
        credits: this.addResource(total.credits, currrentValue.credits),
        knowledge: this.addResource(total.knowledge, currrentValue.knowledge),
        ore: this.addResource(total.ore, currrentValue.ore),
        powerNodes: this.addResource(
          total.powerNodes,
          currrentValue.powerNodes,
        ),
        qic: this.addResource(total.qic, currrentValue.qic),
      },
    };
  }

  private addResource(
    currentValue: number | undefined,
    addition: number | undefined,
  ): number | undefined {
    if (addition) {
      return currentValue ? currentValue + addition : addition;
    }

    return currentValue;
  }

  toJSON(): SerializedState {
    return {
      type: GameState.IncomePhase,
    };
  }
}
