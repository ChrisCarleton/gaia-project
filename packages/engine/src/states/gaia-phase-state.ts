import { SerializedState } from '../core/serialization';
import { ObserverPublisher } from '../events';
import { ChangeStateFunction, GameContext, GameState } from '../interfaces';
import { ActionPhaseState } from './action-phase-state';
import { StateBase } from './state-base';

export class GaiaPhaseState extends StateBase {
  constructor(
    context: GameContext,
    events: ObserverPublisher,
    changeState: ChangeStateFunction,
  ) {
    super(context, events, changeState);
  }

  get currentState(): GameState {
    return GameState.GaiaPhase;
  }

  init(): void {
    // TODO: Convert Transdim planets to Gaia planets and restore power tokens to the power cycle.

    this.changeState(
      new ActionPhaseState(
        this.context,
        this.events,
        this.changeState,
        this.context.players[0],
      ),
    );
  }

  toJSON(): SerializedState {
    return { type: GameState.GaiaPhase };
  }
}
