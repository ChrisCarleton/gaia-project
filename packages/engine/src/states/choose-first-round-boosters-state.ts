import {
  ChangeStateFunction,
  EventType,
  GameAction,
  GameContext,
  GameState,
  ObserverPublisher,
  RoundBooster,
} from '..';
import { SerializedState } from '../core/serialization';
import { IncomePhaseState } from './income-phase-state';
import { StateBase } from './state-base';

export class ChooseFirstRoundBoostersState extends StateBase {
  constructor(
    context: GameContext,
    events: ObserverPublisher,
    changeState: ChangeStateFunction,
    private readonly player: number,
  ) {
    super(context, events, changeState);
  }

  get currentState(): GameState {
    return GameState.ChooseFirstRoundBoosters;
  }

  init(): void {
    this.events.publish({
      type: EventType.AwaitingPlayerInput,
      gameState: GameState.ChooseFirstRoundBoosters,
      allowedActions: [GameAction.SelectRoundBooster],
      player: this.context.players[this.player],
    });
  }

  chooseRoundBoosterAndPass(roundBooster: RoundBooster): void {
    const player = this.context.players[this.player];

    const index = this.context.roundBoosters.findIndex(
      (rb) => rb.id === roundBooster.id,
    );
    if (index === -1) {
      throw new Error('Unknown round booster.');
    }

    this.context.roundBoosters.splice(index, 1);
    this.events.publish({
      type: EventType.RoundBoosterSelected,
      player,
      roundBooster,
    });

    if (this.player === 0) {
      // All done. Advance the game to the next phase.
      this.changeState(
        new IncomePhaseState(this.context, this.events, this.changeState),
      );
    } else {
      this.changeState(
        new ChooseFirstRoundBoostersState(
          this.context,
          this.events,
          this.changeState,
          this.player - 1,
        ),
      );
    }
  }

  toJSON(): SerializedState {
    return {
      type: GameState.ChooseFirstRoundBoosters,
      player: this.player,
    };
  }
}
