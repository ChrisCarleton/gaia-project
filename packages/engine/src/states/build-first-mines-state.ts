import { SerializedState } from '../core/serialization';
import { ErrorCode, GPError } from '../errors';
import { EventType } from '../events/event-args';
import { Observer } from '../events/observer';
import { FactionType } from '../factions';
import {
  ChangeStateFunction,
  GameAction,
  GameContext,
  GameState,
  MapHex,
  Player,
} from '../interfaces';
import { ChooseFirstRoundBoostersState } from './choose-first-round-boosters-state';
import { StateBase } from './state-base';

export enum BuildFirstMinesPass {
  First,
  Second,
  Xenos, // Xenos get to build a third mine after everyone else has built two.
  Ivits, // Ivits do not build mines. Instead, they wait until everyone else has gone and then build their planetary institute.
}
export type BuildFirstMinesOptions = {
  turnIndex: number;
  pass: BuildFirstMinesPass;
};

type DetermineNextPlayerStrategyResult = {
  nextPlayerIndex: number;
  changePass?: BuildFirstMinesPass;
} | null;
type DetermineNextPlayerStrategy = (
  players: Readonly<Player[]>,
  currentIndex: number,
) => DetermineNextPlayerStrategyResult;
const DetermineNextPlayer: Record<
  BuildFirstMinesPass,
  DetermineNextPlayerStrategy
> = {
  // First pass: Iterate over players in order, skipping Ivits.
  [BuildFirstMinesPass.First]: (players, currentIndex) => {
    do {
      currentIndex++;

      if (currentIndex >= players.length)
        return {
          nextPlayerIndex: players.length - 1,
          changePass: BuildFirstMinesPass.Second,
        };
    } while (players[currentIndex].faction.factionType === FactionType.Ivits);

    return { nextPlayerIndex: currentIndex };
  },

  // Second pass: Iterate over players in reverse order, skipping Ivits.
  [BuildFirstMinesPass.Second]: (players, currentIndex) => {
    do {
      currentIndex--;

      if (currentIndex < 0) {
        // Is someone playing as Xenos?
        const xenosIndex = players.findIndex(
          (player) => player.faction.factionType === FactionType.Xenos,
        );
        if (xenosIndex > -1) {
          return {
            nextPlayerIndex: xenosIndex,
            changePass: BuildFirstMinesPass.Xenos,
          };
        }

        // What about Ivits?
        const ivitsIndex = players.findIndex(
          (player) => player.faction.factionType === FactionType.Ivits,
        );
        if (ivitsIndex > -1) {
          return {
            nextPlayerIndex: ivitsIndex,
            changePass: BuildFirstMinesPass.Ivits,
          };
        }

        // No? Cool! We're done.
        return null;
      }
    } while (players[currentIndex].faction.factionType === FactionType.Ivits);

    return { nextPlayerIndex: currentIndex };
  },

  // Xenos get to build a third mine after everyone else has gone.
  [BuildFirstMinesPass.Xenos]: (players) => {
    // Once the Xenos have built their third mine then we just need to check for Ivits.
    const ivitsIndex = players.findIndex(
      (player) => player.faction.factionType === FactionType.Ivits,
    );
    if (ivitsIndex > -1) {
      return {
        nextPlayerIndex: ivitsIndex,
        changePass: BuildFirstMinesPass.Ivits,
      };
    }

    // Otherwise, we're done.
    return null;
  },

  // Ivits go at the very end and build a Planetary Institution.
  [BuildFirstMinesPass.Ivits]: () => {
    // Nothing left to do after the Ivits have gone.
    return null;
  },
};

export class BuildFirstMinesState extends StateBase {
  constructor(
    context: GameContext,
    events: Observer,
    changeState: ChangeStateFunction,
    private readonly options: BuildFirstMinesOptions,
  ) {
    super(context, events, changeState);
  }

  get currentState(): GameState {
    return GameState.BuildFirstMines;
  }

  init(): void {
    // Alert the current player that they need to place a fresh mine.
    this.events.publish({
      type: EventType.AwaitingPlayerInput,
      player: this.context.players[this.options.turnIndex],
      gameState: this.currentState,
      allowedActions: [GameAction.BuildMine],
    });
  }

  buildMine(location: MapHex): void {
    // Identify the player. Unlike normal turn order, initial mine selection is done twice:
    // Once in turn order, then again in reverse turn order.
    // ... and the Xenos get to place a third mine after all that!
    const player = this.context.players[this.options.turnIndex];

    // Validate hex. Must be unoccupied and contain a planet matching the player's home world.
    if (!location.planet) {
      throw new GPError(
        ErrorCode.InvalidMinePlacement,
        'New mine must be built on a planet.',
      );
    }

    if (location.planet.type !== player.faction.homeWorld) {
      throw new GPError(
        ErrorCode.InvalidMinePlacement,
        "Mine can only be built on the player's homeworld type.",
      );
    }

    if (location.planet.structure) {
      throw new GPError(
        ErrorCode.InvalidMinePlacement,
        'Mine can only be placed on an unoccupied planet.',
      );
    }

    // Place mine event.
    this.events.publish({
      type: EventType.MineBuilt,
      player,
      location,
    });

    const nextPlayer = DetermineNextPlayer[this.options.pass](
      this.context.players,
      this.options.turnIndex,
    );

    if (nextPlayer) {
      this.changeState(
        new BuildFirstMinesState(this.context, this.events, this.changeState, {
          pass: nextPlayer.changePass ?? this.options.pass,
          turnIndex: nextPlayer.nextPlayerIndex,
        }),
      );
    } else {
      this.changeState(
        new ChooseFirstRoundBoostersState(
          this.context,
          this.events,
          this.changeState,
          this.context.players[this.context.players.length - 1],
        ),
      );
    }
  }

  toJSON(): SerializedState {
    return {
      type: GameState.BuildFirstMines,
      player: this.options.turnIndex,
      pass: this.options.pass,
    };
  }
}
