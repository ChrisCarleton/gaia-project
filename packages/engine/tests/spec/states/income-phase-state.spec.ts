import { mockDeep } from 'jest-mock-extended';

import {
  BasicMapModel,
  EventType,
  FactionIncome,
  FreeAction,
  GameContext,
  GameState,
  Income,
  ResearchArea,
  ResearchBoard,
  ResearchProgress,
  RoundBooster,
  RoundBoosterBonusType,
  RoundBoosterPassBonusDiscriminator,
  StructureType,
} from '../../../src';
import { GaiaPhaseState } from '../../../src/states/gaia-phase-state';
import { IncomePhaseState } from '../../../src/states/income-phase-state';
import { TestObserver, createTestPlayer } from '../../util';

const events = new TestObserver();
const map = new BasicMapModel().createMap(2);

type PlayerOptions = {
  research: Partial<Record<ResearchArea, number>>;
  roundBooster: RoundBooster;
  structures: Partial<Record<StructureType, number>>;
  structureIncome: Partial<FactionIncome>;
};

const NullStructureIncome: FactionIncome = {
  [StructureType.Mine]: new Array<Income>(9).fill({}),
  [StructureType.TradingStation]: new Array<Income>(5).fill({}),
  [StructureType.PlanetaryInstitute]: [{}, {}],
  [StructureType.ResearchLab]: new Array<Income>(4).fill({}),
};

describe('Income Phase State', () => {
  function createContext(options?: Partial<PlayerOptions>): GameContext {
    const players = [
      createTestPlayer({
        id: '0',
        name: 'Test Player',
        research: options?.research,
        roundBooster: options?.roundBooster,
        structures: options?.structures,
        structureIncome: options?.structureIncome,
      }),
    ];

    return {
      allowedActions: [],
      currentRound: 1,
      currentPlayer: players[0],
      map,
      players,
      researchBoard: mockDeep<ResearchBoard>(),
      roundBoosters: [],
      rounds: [],
    };
  }

  afterEach(() => {
    events.reset();
  });

  it('will return the correct current state', () => {
    const state = new IncomePhaseState(createContext(), events, jest.fn());
    expect(state.currentState).toBe(GameState.IncomePhase);
  });

  it('will award income from established structures', () => {
    const context = createContext({
      structures: {
        mine: 4,
        tradingStation: 3,
        researchLab: 2,
        planetaryInstitute: 1,
      },
    });
    const changeState = jest.fn();
    const state = new IncomePhaseState(context, events, changeState);

    state.init();

    expect(events.events).toHaveLength(1);
    const [args] = events.events;
    expect(args.type).toBe(EventType.IncomeGained);
    if (args.type === EventType.IncomeGained) {
      expect(args.player).toBe(context.players[0]);
      expect(args.income).toMatchSnapshot();
    }
    expect(changeState).toBeCalledTimes(1);
  });

  const roundBoosters: Record<string, RoundBooster> = {
    'no income': {
      id: 0,
      a: {
        type: RoundBoosterBonusType.BonusOnPass,
        discriminator: RoundBoosterPassBonusDiscriminator.Mines,
        vp: 2,
      },
      b: {
        type: RoundBoosterBonusType.Action,
        action: FreeAction.BuildMineOrStartGaiaWithRangeBoost,
      },
    },
    'single income': {
      id: 1,
      a: {
        type: RoundBoosterBonusType.BonusOnPass,
        discriminator:
          RoundBoosterPassBonusDiscriminator.PlanetaryInstitutesAndAcadamies,
        vp: 4,
      },
      b: {
        type: RoundBoosterBonusType.Income,
        income: {
          qic: 1,
          powerNodes: 2,
        },
      },
    },
    'dual income': {
      id: 2,
      a: {
        type: RoundBoosterBonusType.Income,
        income: {
          ore: 7,
          knowledge: 2,
        },
      },
      b: {
        type: RoundBoosterBonusType.Income,
        income: {
          credits: 4,
          powerNodes: 2,
        },
      },
    },
  };

  Object.entries(roundBoosters).forEach(([test, roundBooster]) => {
    it(`will award income from round boosters: ${test}`, () => {
      const context = createContext({
        roundBooster,
        structureIncome: NullStructureIncome,
      });
      const changeState = jest.fn();
      const state = new IncomePhaseState(context, events, changeState);

      state.init();

      expect(events.events).toHaveLength(1);
      const [args] = events.events;
      expect(args.type).toBe(EventType.IncomeGained);
      if (args.type === EventType.IncomeGained) {
        expect(args.player).toBe(context.players[0]);
        expect(args.income).toMatchSnapshot();
      }
      expect(changeState).toBeCalledTimes(1);
    });
  });

  const ResearchVariations: Partial<ResearchProgress>[] = [
    { economics: 0, science: 0 },
    { economics: 0, science: 2 },
    { economics: 0, science: 4 },
    { economics: 0, science: 5 },
    { economics: 1, science: 0 },
    { economics: 3, science: 0 },
    { economics: 5, science: 0 },
    { economics: 2, science: 3 },
    { economics: 4, science: 2 },
  ];

  ResearchVariations.forEach((research) => {
    it(`will award income from research progress: Economics ${research.economics}, Science ${research.science}`, () => {
      const context = createContext({
        research,
        structureIncome: NullStructureIncome,
      });
      const changeState = jest.fn();
      const state = new IncomePhaseState(context, events, changeState);

      state.init();

      expect(events.events).toHaveLength(1);
      const [args] = events.events;
      expect(args.type).toBe(EventType.IncomeGained);
      if (args.type === EventType.IncomeGained) {
        expect(args.player).toBe(context.players[0]);
        expect(args.income).toMatchSnapshot();
      }
    });
  });

  it.todo('Test income from Tech Tiles');

  it.todo('will combine all income correctly');

  it('will advance to the first Gaia phase when done', () => {
    const context = createContext();
    const changeState = jest.fn();
    const state = new IncomePhaseState(context, events, changeState);

    state.init();

    expect(changeState).toBeCalledTimes(1);
    expect(changeState.mock.lastCall![0]).toBeInstanceOf(GaiaPhaseState);
  });
});
