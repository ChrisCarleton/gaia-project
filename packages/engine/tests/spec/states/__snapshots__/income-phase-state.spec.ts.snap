// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Income Phase State will award income from established structures 1`] = `
{
  "chargePower": 4,
  "credits": 11,
  "knowledge": 3,
  "ore": 4,
  "powerNodes": 1,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from research progress: Economics 0, Science 0 1`] = `
{
  "chargePower": undefined,
  "credits": undefined,
  "knowledge": undefined,
  "ore": undefined,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from research progress: Economics 0, Science 2 1`] = `
{
  "chargePower": undefined,
  "credits": undefined,
  "knowledge": 2,
  "ore": undefined,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from research progress: Economics 0, Science 4 1`] = `
{
  "chargePower": undefined,
  "credits": undefined,
  "knowledge": 4,
  "ore": undefined,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from research progress: Economics 0, Science 5 1`] = `
{
  "chargePower": undefined,
  "credits": undefined,
  "knowledge": undefined,
  "ore": undefined,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from research progress: Economics 1, Science 0 1`] = `
{
  "chargePower": 1,
  "credits": 2,
  "knowledge": undefined,
  "ore": undefined,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from research progress: Economics 2, Science 3 1`] = `
{
  "chargePower": 2,
  "credits": 2,
  "knowledge": 3,
  "ore": 1,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from research progress: Economics 3, Science 0 1`] = `
{
  "chargePower": 3,
  "credits": 3,
  "knowledge": undefined,
  "ore": 1,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from research progress: Economics 4, Science 2 1`] = `
{
  "chargePower": 4,
  "credits": 4,
  "knowledge": 2,
  "ore": 2,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from research progress: Economics 5, Science 0 1`] = `
{
  "chargePower": undefined,
  "credits": undefined,
  "knowledge": undefined,
  "ore": undefined,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from round boosters: dual income 1`] = `
{
  "chargePower": undefined,
  "credits": 4,
  "knowledge": 2,
  "ore": 7,
  "powerNodes": 2,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from round boosters: no income 1`] = `
{
  "chargePower": undefined,
  "credits": undefined,
  "knowledge": undefined,
  "ore": undefined,
  "powerNodes": undefined,
  "qic": undefined,
}
`;

exports[`Income Phase State will award income from round boosters: single income 1`] = `
{
  "chargePower": undefined,
  "credits": undefined,
  "knowledge": undefined,
  "ore": undefined,
  "powerNodes": 2,
  "qic": 1,
}
`;
